subprojects/docs/src/docs/userguide/custom_gradle_types.adoc

```diff
@@ -0,0 +1,272 @@
+====
+
+==== Mutable managed properties
+
+To declare a mutable managed property, add an abstract getter method and an abstract setter method for the property to the type.
+
+Here is an example of a task type with a `uri` property:
+
+.Mutable managed property
+====
+[source.multi-language-sample,java]
+.UrlProcess.java
+----
+include::{samplesPath}/userguide/plugins/mutableManagedProperty/java/buildSrc/src/main/java/UrlProcess.java[]
+----
+[source.multi-language-sample,kotlin]
+.UrlProcess.kt
+----
+include::{samplesPath}/userguide/plugins/mutableManagedProperty/kotlin/buildSrc/src/main/kotlin/UrlProcess.kt[]
+----
+[source.multi-language-sample,groovy]
+.UrlProcess.groovy
+----
+include::{samplesPath}/userguide/plugins/mutableManagedProperty/groovy/buildSrc/src/main/groovy/UrlProcess.groovy[]
+----
+====
+
+Note that for a property to be considered a mutable managed property, _all_ of the property's getter methods and setter methods must be `public` or `protected` and `abstract`.
+
+==== Read-only managed properties
+
+To declare a read-only managed property, add an abstract getter method for the property to the type.
+The property should not have any setter methods.
+This is a useful pattern to use with one of Gradle's configurable <<lazy_configuration.adoc#lazy_configuration,lazy property>> types.
+
+Here is an example of a task type with a `uri` property:
+
+.Read-only managed property
+====
+[source.multi-language-sample,java]
+.UrlProcess.java
+----
+include::{samplesPath}/userguide/plugins/readOnlyManagedProperty/java/buildSrc/src/main/java/UrlProcess.java[]
+----
+[source.multi-language-sample,kotlin]
+.UrlProcess.kt
+----
+include::{samplesPath}/userguide/plugins/readOnlyManagedProperty/kotlin/buildSrc/src/main/kotlin/UrlProcess.kt[]
+----
+[source.multi-language-sample,groovy]
+.UrlProcess.groovy
+----
+include::{samplesPath}/userguide/plugins/readOnlyManagedProperty/groovy/buildSrc/src/main/groovy/UrlProcess.groovy[]
+----
+====
+
+Note that for a property to be considered a read only managed property, _all_ of the property's getter methods must be `public` or `protected` and `abstract` and the property must not have any setter methods.
+In addition, the property must have one of the following types:
+
+- `Property`
+- `RegularFileProperty`
+- `DirectoryProperty`
+- `ListProperty`
+- `SetProperty`
+- `MapProperty`
+- `ConfigurableFileCollection`
+
+[[managed_types]]
+=== Managed types
+
+A _managed type_ is an abstract class or interface with no fields and whose properties are all managed.
+That is, it is a type whose state is entirely managed by Gradle.
+
+== DSL support and extensibility
+
+When Gradle creates an instance of a custom type, it _decorates_ the instance to mix-in DSL and extensibility support.
+
+Each decorated instance implements link:{javadocPath}/org/gradle/api/plugins/ExtensionAware.html[ExtensionAware], and so can have extension
+objects attached to it.
```

Comment:
❓ the runtime decoration only makes the contract available to non-static Groovy plugins and Groovy scripts. If I didn't miss anything, in order to make it available to statically compiled code, Java or Kotlin plugins and Kotlin scripts, your DSL type can be abstract and implement `ExtensionAware`, then Gradle provides the implementation and your type statically exposes that it can have extensions. I think this should be covered at some point. Maybe we want to add support for non-abstract DSL types too before documenting that, maybe via some `objectFactory.extensionsContainer()`.

Reply:
Right. I intentionally didn't go into the topic of how the DSL and extensibility works too much, as it's not really the purpose of this PR. I just wanted to add a placeholder to make it clearer that we're missing a bunch of detail about how the DSL works (or it's spread across a bunch of places).

wrt statically declaring `ExtensionAware`, this isn't part of the public API yet so I don't want to document it until we've cleaned up the public contract.